### PR TITLE
Suppress and revive offers based on instance state

### PR DESCRIPTION
**Pull request by meichstedt**
Summary:
Changed the ReviveOffersActor to react to instance state changes, instead of being triggered from within the launch queue. Every time an instance is scheduled, a reviveOffers call will be issued. Once all scheduled instances transitioned to another state, offers will be suppressed.

It is to be discussed if and how we want to handle revives when Marathon needs to unreserve/destroy. Also, we should not revive when scheduled instances are implicitly backed off.

JIRA issues: MARATHON-8627

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/core/flow/FlowModule.scala [hunk 1]
```diff
@@ -29,16 +25,15 @@ class FlowModule(leadershipModule: LeadershipModule) extends StrictLogging {
...
     marathonEventStream: EventStream,
-    offersWanted: Source[Boolean, Cancellable],
     driverHolder: MarathonSchedulerDriverHolder): Option[OfferReviver] = {
 
     if (conf.reviveOffersForNewApps()) {
       lazy val reviveOffersActor = ReviveOffersActor.props(
-        metrics, clock, conf, marathonEventStream,
-        offersWanted, driverHolder
+        metrics, conf, marathonEventStream,
+        ???,
```
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/core/flow/FlowModule.scala]: This is missing the instanceTracker

File: src/main/scala/mesosphere/marathon/core/launchqueue/impl/ReviveOffersActor.scala [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    case Status.Failure(cause) =>
+      // escalate this failure
+      throw new IllegalStateException("while loading instances", cause)
+
+    case _: AnyRef =>
+      stash()
+  }
+
+  // state that reacts to instance changes
+  def initialized(scheduledInstances: HashSet[Instance.Id]): Receive = LoggingReceive {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/launchqueue/impl/ReviveOffersActor.scala]: This is such a nice state machine. I really like it.

File: src/main/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActor.scala [hunk 3]
```diff
@@ -204,19 +200,6 @@ private class TaskLauncherActor(
...
-        // we already rejected. E.g. when a host:unique constraint prevented
-        // us to launch tasks on a particular node before, we need to reconsider offers
-        // of that node after a task on that node has died.
-        //
-        // B) If a reservation timed out, already rejected offers might become eligible for creating new reservations.
-        val runSpec = update.instance.runSpec
-        if (runSpec.constraints.nonEmpty || (runSpec.isResident && shouldLaunchInstances(clock.now()))) {
-          maybeOfferReviver.foreach(_.reviveOffers())
-        }
-      }
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActor.scala]: 🎉  @alenkacz

File: src/main/scala/mesosphere/marathon/core/matcher/reconcile/impl/OffersWantedForReconciliationActor.scala [hunk 4]
```diff
@@ -7,13 +7,14 @@ import java.time.Clock
...
 import com.typesafe.scalalogging.StrictLogging
-import mesosphere.marathon.core.flow.ReviveOffersConfig
 import mesosphere.marathon.core.event.DeploymentStepSuccess
 import mesosphere.marathon.state.Timestamp
 import mesosphere.marathon.core.deployment.StopApplication
+import mesosphere.marathon.core.launchqueue.ReviveOffersConfig
 
 import scala.concurrent.duration._
 
+// TODO: this actor works against the ReviveOffersActor and needs to be adjusted/removed
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/core/matcher/reconcile/impl/OffersWantedForReconciliationActor.scala]: If I'm not mistaken we need the revives for the `OfferMatcherReconciler` which should be refactored anyways for [MARATHON-8517](https://jira.mesosphere.com/browse/MARATHON-8517).

When is this actor triggered? Maybe be can fold the logic into the new `ReviveOffersActor`.

File: src/test/scala/mesosphere/marathon/core/launchqueue/impl/ReviveOffersActorTest.scala [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+      val holder = new MarathonSchedulerDriverHolder
+      holder.driver = Some(driver)
+      holder
+    }
+    val instanceTracker: InstanceTracker = mock[InstanceTracker]
+    instanceTracker.instanceUpdates returns Source.empty // TODO: needed?
+
+    val metrics: Metrics = DummyMetrics
+
+    lazy val actorRef: TestActorRef[ReviveOffersActor] = TestActorRef[ReviveOffersActor](
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/core/launchqueue/impl/ReviveOffersActorTest.scala]: Why not make the logic a simple trait or class such as in `ExpungeOverdueLostTasksActorLogic`. This wy we don't need to test the actor but a simple class which is much faster and simpler. No racy tests!

File: src/main/scala/mesosphere/marathon/core/launchqueue/ReviveOffersConfig.scala [hunk 6]
```diff
@@ -1,10 +1,11 @@
 package mesosphere.marathon
-package core.flow
+package core.launchqueue
 
 import org.rogach.scallop.ScallopConf
 
 trait ReviveOffersConfig extends ScallopConf {
 
+  // TODO: deprecate this
```
Comment (timcharper) [inline src/main/scala/mesosphere/marathon/core/launchqueue/ReviveOffersConfig.scala]: We should probably have a JIRA for these deprecations

File: src/main/scala/mesosphere/marathon/core/launchqueue/impl/ReviveOffersActor.scala [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+  private[this] val suppressCountMetric: Counter = metrics.counter("mesos.calls.suppress")
+
+  override def preStart(): Unit = {
+    super.preStart()
+
+    import akka.pattern.pipe
+    import context.dispatcher
+    instanceTracker.instancesBySpec().pipeTo(self)
+
+    eventStream.subscribe(self, classOf[InstanceChanged])
```
Comment (timcharper) [inline src/main/scala/mesosphere/marathon/core/launchqueue/impl/ReviveOffersActor.scala]: Let's use the instance tracker subscription mechanism to ensure that we have consistent replication of state.

**Comment by jeschkies** [comment]
Superseded by 6936

**Pull request closed without merging**